File: Java/127. Word Ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Map<String, List<String>> transform = new HashMap<>();
        int L = beginWord.length();
        
        wordList.forEach( word -> {
            for(int i=0; i<L; i++){
                String cur = word.substring(0, i) + '*' + word.substring(i+1, L);
                List<String> trans = transform.getOrDefault(cur, new ArrayList<>());
                trans.add(word);
                transform.put(cur, trans);
            }
        });
        
        Queue<Pair<String, Integer>> q = new LinkedList<>();
        q.add(new Pair(beginWord, 1));
        Set<String> visited = new HashSet<>();
        visited.add(beginWord);
        
        while(!q.isEmpty()) {
            Pair<String, Integer> curr = q.poll();
            String w = curr.getKey();
            int dis  = curr.getValue();
            for(int i=0; i<L; i++) {
                String trans = w.substring(0, i) + '*' + w.substring(i+1, L);
                for(String nei: transform.getOrDefault(trans, new ArrayList<>())) {
                    if(nei.equals(endWord)) {
                        return dis+1;
                    }
                    
                    if(!visited.contains(nei)) {
                        visited.add(nei);
                        q.offer(new Pair(nei, dis+1));
                    }
                }
            }
        }
        return 0;
    }
}

File: Python3/127. Word Ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        transform = collections.defaultdict(list)
        
        # all words have the same length
        L = len(beginWord)
        
        # group up words which have only one different letter
        for w in wordList:
            for i in range(L):
                transform[w[:i] + '*' + w[i+1:]].append(w)
                
        # use a queue to perform BFS
        # and a set to mark visited words
        q = deque([(beginWord, 1)])
        visited = set(beginWord)
        
        while q:
            cur, dis = q.popleft()
            for i in range(L):
                trans = cur[:i] + '*' + cur[i+1:]
                for w in transform[trans]:
                    # goal reached
                    if w == endWord:
                        return dis+1
                    
                    # otherwise, add it to the queue
                    if w not in visited:
                        q.append((w, dis+1))
                        visited.add(w)
                # mark this transformation as null so we won't check again
                transform[trans] = []
                
        return 0
                